cairo/ethereum/cancun/state.cairo

```diff
@@ -210,6 +211,65 @@ func get_account_optional{poseidon_ptr: PoseidonBuiltin*, state: State}(
     return account;
 }
 
+// @notice Returns the code for the given account.
+// If the code is not cached, it is loaded from the program input. The program
+// input must contain a Dict[Tuple[Low, High], Bytes] where the key is the codehash
+// (in its cairo representation) and the value is the code.
+// @dev: Accesses to the `code` field __MUST__ be done using `get_account_code`.
+func get_account_code{
+    range_check_ptr,
+    bitwise_ptr: BitwiseBuiltin*,
+    keccak_ptr: KeccakBuiltin*,
+    poseidon_ptr: PoseidonBuiltin*,
+    state: State,
+}(address: Address, account: Account) -> Bytes {
+    alloc_locals;
+
+    // Account code is already cached - return it
+    if (cast(account.value.code.value, felt) != 0) {
+        return account.value.code;
+    }
+
+    // Account code is not cached - load it and ensure hash(code) == code_hash
+    // TODO: this should not be triggered in tests - only in prod script.
+    // TODO: ensure this works with a test.
+    tempvar code: felt*;
+    tempvar code_len: felt;
+    %{
+        # TODO: do we want to serialize (I fear it might be slow) here?
+        # best would be for the program input to already have the key:Tuple[Low, High] -> code
+        account_code = program_input["codehash_to_code"][serialize(ids.account.value.code_hash)];
+        segment.load_data(code, account_code);
+        code_len = len(account_code);
+    %}
+    tempvar account_code = Bytes(new BytesStruct(data=code, len=code_len));
+
+    // Soundness checks: ensure that hash(account_code) == account.value.code_hash
+    let code_hash = keccak256(account_code);
+    with_attr error_message("AssertionError") {
```

Comment:
are we not using INVARIANT anymore?
feel free to change other invariant calls to become AssertionErrors

Reply:
doesn't really matter, by convention we try to map python types, but if there's no type to match (e.g. here we don't expect to mirror a pythonic behavior we could use something more verbose)